analyzer.py

```diff
@@ -0,0 +1,83 @@
+import math
+import numpy as np
+import pandas as pd
+import scipy as cp
+
+DAYS_PER_YEAR = 365
+
+class Analyzer():
+    """
+    This class calculates and stores the dividends_by_year DataFram, the DGR for the given number of years (dgr_years)
+    """
+    def __init__(self, dividends_df: pd.core.series.Series, dgr_years: float) -> None:
+        self.dividends_df = dividends_df
+        self.dgr_years = dgr_years
+        
+        self._dividends_by_year = pd.DataFrame({"Year": [], "Dividends": [], "Count": []})
+        self._dgr = math.nan
+        self._growth_streak = math.nan
+        
+        if self.dividends_df.size > 2:
+            self.cal_yearly_dividends()
+            self.cal_dgr()
+            self.cal_growth_streak()
+    
+    def get_compact_data(self) -> dict:
+        return {
+            "DGR": self._dgr,
+            "growth_streak": self._growth_streak,
+        }
+
+    @property
+    def dividends_by_year(self) -> pd.DataFrame:
+        return self._dividends_by_year
+    
+    @property
+    def dgr(self) -> float:
+        return self._dgr
+    
+    @property
+    def growth_streak(self) -> float:
+        return self._growth_streak
+    
+
+    def cal_yearly_dividends(self):
+        """
+        This function calculated the dividends_by_year DataFrame, which shows how many dividends were given in each year since the company started to share dividends
+        """
+        years = self.dividends_df["Date"].values.astype("datetime64[Y]").astype(int)+1970
+        firt_year = np.min(years)
+        last_year = np.datetime64('now').astype("datetime64[Y]").astype(int)+1970
+        divs = self.dividends_df["Dividends"].values
+        for year in range(firt_year, last_year+1):
+            inds = years == year
+            yearly_div = np.sum(divs[inds])
+            count = inds.sum()
+            self._dividends_by_year.loc[len(self._dividends_by_year.index)] = [year, yearly_div, count]    
+
+    def cal_dgr(self):
+        """
+        This function fits the data from the last number_of_years to an exponential model and calculates dividend growth rate (DGR)
+        To avoid negative biases the function discluding the current year
+        """
```

Comment:
when I run this function the code crashes
`
Traceback (most recent call last):
  File "/home/dorsadeh/workspace/sandbox/python_scripts/personal/finance/main.py", line 106, in <module>
    get_data(metrics, tickers, settings.years_dividends_growth)
  File "/home/dorsadeh/workspace/sandbox/python_scripts/personal/finance/main.py", line 36, in get_data
    div_analysis = analyzer.Analyzer(dividends_df, years_dividends_growth)
  File "/home/dorsadeh/workspace/sandbox/python_scripts/personal/finance/analyzer.py", line 22, in __init__
    self.cal_dgr()
  File "/home/dorsadeh/workspace/sandbox/python_scripts/personal/finance/analyzer.py", line 63, in cal_dgr
    divs = self._dividends_by_year["Dividends"].values[-self.dgr_years-1:-1]
TypeError: slice indices must be integers or None or have an __index__ method
`

The dividend data is 
Date,Dividends
2008-08-13 00:00:00-04:00,0.2
2008-11-14 00:00:00-05:00,0.2
2009-02-13 00:00:00-05:00,0.2
2009-05-14 00:00:00-04:00,0.2
2009-08-14 00:00:00-04:00,0.21
2009-11-16 00:00:00-05:00,0.21
2010-02-16 00:00:00-05:00,0.21
2010-05-14 00:00:00-04:00,0.21
2010-08-16 00:00:00-04:00,0.22
2010-11-16 00:00:00-05:00,0.22
2011-02-16 00:00:00-05:00,0.22
2011-05-16 00:00:00-04:00,0.22
2011-08-10 00:00:00-04:00,0.23
2011-11-16 00:00:00-05:00,0.23
2012-02-01 00:00:00-05:00,0.23
2012-04-18 00:00:00-04:00,0.23
2012-07-03 00:00:00-04:00,0.25
2012-11-14 00:00:00-05:00,0.25
2012-12-18 00:00:00-05:00,0.25
2013-05-22 00:00:00-04:00,0.28
2013-08-15 00:00:00-04:00,0.28
2013-11-13 00:00:00-05:00,0.28
2014-01-30 00:00:00-05:00,0.28
2014-05-08 00:00:00-04:00,0.31
2014-08-07 00:00:00-04:00,0.31
2014-11-06 00:00:00-05:00,0.31
2015-02-05 00:00:00-05:00,0.31
2015-05-07 00:00:00-04:00,0.34
2015-08-06 00:00:00-04:00,0.34
2015-11-05 00:00:00-05:00,0.34
2016-02-04 00:00:00-05:00,0.34
2016-05-05 00:00:00-04:00,0.375
2016-08-04 00:00:00-04:00,0.375
2016-11-07 00:00:00-05:00,0.375
2017-02-03 00:00:00-05:00,0.375
2017-05-03 00:00:00-04:00,0.415
2017-08-07 00:00:00-04:00,0.415
2017-11-09 00:00:00-05:00,0.415
2018-02-06 00:00:00-05:00,0.415
2018-05-10 00:00:00-04:00,0.455
2018-08-09 00:00:00-04:00,0.455
2018-11-08 00:00:00-05:00,0.455
2019-02-06 00:00:00-05:00,0.455
2019-05-10 00:00:00-04:00,0.5
2019-08-08 00:00:00-04:00,0.5
2019-11-08 00:00:00-05:00,0.5
2020-02-06 00:00:00-05:00,0.5
2020-05-11 00:00:00-04:00,0.55
2020-08-11 00:00:00-04:00,0.55
2020-11-09 00:00:00-05:00,0.55
2021-02-05 00:00:00-05:00,0.55
2021-05-10 00:00:00-04:00,0.603
2021-08-09 00:00:00-04:00,0.603
2021-11-09 00:00:00-05:00,0.603
2022-02-07 00:00:00-05:00,0.603
2022-05-09 00:00:00-04:00,0.655
2022-08-08 00:00:00-04:00,0.655
2022-11-07 00:00:00-05:00,0.655
2023-02-06 00:00:00-05:00,0.655

Reply:
This is not a bug. The datetime of this particular ticker (and maby more in your repository) is not updated with the correct format. Just refresh the data and it should work.